### DISCUSSION TITLE
Disabling Next.js server side rendering to prevent window is undefined errors, or other solutions.

### DISCUSSION THREAD

**Discussion by damiangene** [vercel/next.js#16327] [Help]
I have a private NPM package that I maintain, which I use in a Next.js project. Both projects are React, Typescript projects.

I've recently added a graph into the NPM project and have now run into the problem of; all the calls to window within that NPM package throw an error `window is not defined` within the Next.js project

As I'm using a library, that I don't have control over, to build my graph, I don't have the luxury of wrapping the calls to window in some sort of `if (typeof window !== "undefined")` statement.

I get these errors on build of my Next.js project, before I've even made a call to a component that makes use of the graphing library. Simply including my NPM package in my main project as a dependency results in these errors arising.

Would it be possible to stop server-side rendering just for that particular graphing library or perhaps my entire NPM package?

Alternatively, any other solution would be welcome, however wild.

**Comment by jamesmosier** [comment]
You can dynamically import React components that are not server rendered using [next/dynamic](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr) with `ssr: false`

```js
const DynamicComponentWithNoSSR = dynamic(
  () => import('../my-component-that-requires-window'),
  { ssr: false }
)
```

This only works with React components, so you could put whatever logic requires `window` into a React component in your project and then dynamically import it with `ssr: false`

**Reply by damiangene** [reply]
This has been suggested before but doesn't resolve the problem I'm having because I get the error just by having the NPM package installed.

**Reply by jamesmosier** [reply]
What's the NPM package in particular? Do you have a sample repo available? I've never heard of this technique *not* working.

**Comment by damiangene** [comment]
Ok so, I fixed the problem seemingly. Still need to do some more testing.

Isolation was the answer. Because I have an index.ts file in my NPM project where I was exporting the components that were causing errors. 

Now I can use the method you originally proposed.

What I'm struggling with now is using the props that exist for the components in question.

```
const myComponent = dynamic(
  () => import('../components/myComponent').then(mod => mod.myComponent),
  { 
    ssr: false ,
  }
)

export default function Home() {
return (
    <div>
      <Head>
        <title>My site</title>
      </Head>
      <main>
          <myComponent name="Damian" age={27} />
      </main>
    </div>
  )
}
```


This is failing with the error 
`Type '{ name: string; age: number; }' is not assignable to type 'IntrinsicAttributes & { children?: ReactNode; }'.`

**Reply by NBruhno** [reply]
To get TS to "properly" resolve the types for my named dynamic imports, I had to do it like this instead:
```tsx
const MyComponent = dynamic(async () => {
  const module = await import('components/MyComponent')
  return module.MyComponent
}, { ssr: false })
```

But there is probably a better way to do it.

**Reply by damiangene** [reply]
No luck for me with this. Ill look into it a little more a bit later.

Its not picking up the properties that my component needs, So I cant currently pass it anything

**Comment by aprilmintacpineda** [comment]
Still no global solution? Damn, it was a mistake to use NextJS. Maybe for a website nextjs is good but for a web app, it's pretty bad.
